### PR TITLE
Implement collections/sortBy ( closes #1061 )

### LINKED ISSUE #1061: Feature request: `sortBy` function in `std/collections`
**Is your feature request related to a problem? Please describe.**

`Array.prototype.sort` sorts an array using a callback that takes two elements of the array and must return a number, whose sign will determine their sort order. This is not very intuitive, especially when you just want to compare the elements by the natural ordering of one of their fields, for example.

A related issue is that, if you don't pass a callback to `Array.prototype.sort`, the default behavior is to stringify the array elements, which is very counterintuitive in cases like `[10, 20, 110].sort()`.

**Describe the solution you'd like**

Add a `sortBy` utility function in `std/collections`, whose callback parameter maps an array element to a value which will then be compared with the mapped values for other array elements, much like the `key` callback parameter to Python's [`list.sort`](https://docs.python.org/3/library/stdtypes.html#list.sort). The mapped values would be compared using their natural ordering (a.k.a. by using the regular comparison operators).

And since (as per above), `Array.prototype.sort` without a callback also has problems, the callback attribute to `sortBy` should also be optional and default to the identity function.

**Describe alternatives you've considered**

An alternative would be to have a function that would take a mapping callback and return a callback that can be used with `Array.prototype.sort`. This would incur additional complexity that probably isn't needed.

**Additional context**

N/A

### PR DESCRIPTION
**Pull request by LionC**
This PR adds a `sortBy` function to the `collections` module.

`sortBy` allows you to sort an array without mutating it by an arbitrary selected value like this:

```ts
const people = [
    { name: 'Anna', age: 34 },
    { name: 'Kim', age: 42 },
    { name: 'John', age: 23 },
]

const sortedByAge = sortBy(people, it => it.age)
```

`sortBy` was part of the original proposal (see #1065 and its links) and in the discussions following the merge of the initial set of functions, seemed to have consensus, which is why I went forward and implemented it.

Notes on the implementation: I chose to limit the selector function to common sortable types. I looked into different options and they all seemed hacky or inconvenient for users of the function. The approach in the PR seemed like the best compromise between not caring about return types at all (and thus not saving people from hard to track bugs) and specifying every possible case.

This closes #1061

### CODE REVIEW

File: collections/sort_by.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+ *
+ * Example:
+ *
+ * ```typescript
+ * const people = [
+ *     { name: 'Anna', age: 34 },
+ *     { name: 'Kim', age: 42 },
+ *     { name: 'John', age: 23 },
+ * ]
+ * const sortedByAge = sortBy(people, it => it.age)
```
Comment (kt3k) [inline collections/sort_by.ts:14]: We set up doc example type checking in #1050 . So this jsdoc example causes an error in CI saying "sortBy is not found". Please add `import { sortBy } from "./sort_by.ts";` at the beginning of the example block.

File: collections/sort_by.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
+// Copyright 2018-2021 the Deno authors. All rights reserved. MIT license.
+
+/**
+ * Returns all elements in the given collection, sorted by their result using the given selector
+ *
+ * Example:
+ *
+ * ```typescript
```
Comment (kt3k) [inline collections/sort_by.ts:8]: We prefer `ts` tag for typescript example block. ref: https://github.com/denoland/deno_std/pull/1050#discussion_r673884120

File: collections/sort_by.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+ * ```
+ */
+export function sortBy<T>(
+  array: Array<T>,
+  selector:
+      | ((el: T) => number)
+      | ((el: T) => string)
+      | ((el: T) => bigint)
+      | ((el: T) => boolean)
+      | ((el: T) => Date),
```
Comment (kt3k) [inline collections/sort_by.ts:30]: I like this idea of restricting the return value :+1: This should reduce the confusion in the user's code base.

But I'm not sure if we should include boolean as returned value. Do you think there is a use case for that? If the purpose of sorting by boolean is to separate items in 2 groups, how about encouraging the use of `partition`?

**Comment by kt3k** [comment]
The test cases look good to me.

File: collections/sort_by.ts (hunk 3 above)
Comment (LionC) [reply collections/sort_by.ts:30]: I have no strong opinion on this - I just included the builtins / primitives that get sorted "correctly" by the spec. I agree that `boolean` might be less useful. If no one objects, I will remove it.

File: collections/sort_by.ts (hunk 1 above)
Comment (LionC) [reply collections/sort_by.ts:14]: Oh the examples are not evaluated in the scope of the current file? I was not aware. That makes the examples kind of verbose doesn't it?

File: collections/sort_by.ts (hunk 2 above)
Comment (LionC) [reply collections/sort_by.ts]: Fixed

File: collections/sort_by.ts (hunk 1 above)
Comment (LionC) [reply collections/sort_by.ts:16]: I added the import - but it still feels like a lot of clutter for an example that is meant to be displayed in an LSP hover :-S

File: collections/sort_by.ts (hunk 3 above)
Comment (LionC) [reply collections/sort_by.ts]: I removed it

**Review by kt3k — APPROVED** [review]
@LionC LGTM. Nice work!

**Pull request merged**